cmake/modules/hunter_download_server_url.cmake

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2017 NeroBurner
+# All rights reserved.
+
+include(CMakeParseArguments) # cmake_parse_arguments
+
+include(hunter_status_print)
+include(hunter_internal_error)
+include(hunter_test_string_not_empty)
+
+# used gobal variables
+# - HUNTER_DOWNLOAD_SERVER (mandatory)
+#   - URLs are updated to point to the download server
+# - HUNTER_DOWNLOAD_SERVER_BLACKLIST
+#   - blacklist specifies server urls, which are not mapped to the download server
+
+# update download url to point to HUNTER_DOWNLOAD_SERVER
+# - PACKAGE:       name of the package to update the url for
+# - version:       version of the package
+# - URL:           original URL for the package
+# - OUTPUT:        variable to output the updated URL
+function(hunter_download_server_url)
+  set(function_name "hunter_download_server_url")
+  set(function_synopsis "${function_name}(PACKAGE foo VERSION 0.1.2-p3 URL http://original_url.example.com OUTPUT variable)")
+
+  # parse arguments
+  set(one_value_args PACKAGE VERSION URL OUTPUT)
+  cmake_parse_arguments(x "" "${one_value_args}" "" ${ARGV})
+
+  # No free arguments allowed
+  list(LENGTH x_UNPARSED_ARGUMENTS x_len)
+  if(NOT x_len EQUAL 0)
+    hunter_internal_error(
+      "'${function_name}' incorrect usage,"
+      " expected no free arguments '${x_UNPARSED_ARGUMENTS}'."
+      " Synopsis: ${function_synopsis}"
+    )
+  endif()
+
+  # check mandatory arguments
+  foreach(arg PACKAGE VERSION URL OUTPUT)
+    string(COMPARE EQUAL "${x_${arg}}" "" is_empty)
+    if(is_empty)
+      hunter_internal_error(
+        "'${function_name}' incorrect usage,"
+        " option '${arg}' with one argument is mandatory."
+        " Synopsis: ${function_synopsis}"
+      )
+    endif()
+  endforeach()
+
+  hunter_test_string_not_empty("${HUNTER_DOWNLOAD_SERVER}")
+
+  # update download url to point to HUNTER_DOWNLOAD_SERVER
+  # - blacklist specifies server urls, which are not mapped to the download server
+  set(pkg_url_in_blacklist FALSE)
+  foreach(blacklist_item ${HUNTER_DOWNLOAD_SERVER_BLACKLIST})
```

Comment:
> HUNTER_DOWNLOAD_SERVER_BLACKLIST

The name is confusing, it's not blacklisting anything.

Also I'm not sure what is the point of this feature. If you're using your own server why not download everything?

Reply:
The reason for this feature is, that I have an internal GitLab server with some private packages. Without this blacklist I would have to create a tag and then copy the file to the download server.

I chose blacklist, because it removes package URLs from the list of URLs to redirect to the download server

Reply:
> Without this blacklist I would have to create a tag and then copy the file to the download server

How do you use `HUNTER_DOWNLOAD_SERVER`? I was thinking that you've collected all source archives and put it to server.

> I chose blacklist, because it removes package URLs from the list of URLs to redirect to the download server

From https://en.oxforddictionaries.com/definition/blacklist :

```
A list of people or things that are regarded as unacceptable or
untrustworthy and should be excluded or avoided
```

`HUNTER_DOWNLOAD_SERVER_BLACKLIST` sounds like list of servers that you don't want to trust. I expect error while trying to download sources from there.

Reply:
hmm, then maybe whitelist would be better? or do you maybe have a better name?

Reply:
> maybe whitelist would be better? 

Sounds like an opposite to blacklist. I don't think it's something about trusted/untrusted.

> do you maybe have a better name?

`NO_REDIRECT`, `ALWAYS_DIRECT`? Not sure, still confusing - unrelated [URL redirection](https://en.wikipedia.org/wiki/URL_redirection) came to mind.

However like I said already I don't fully understand the use case. Can you add few details?

Reply:
I've updated it to `HUNTER_DOWNLOAD_SERVER_NO_REDIRECT`

My build environment has no internet access. I intend to put the archives of external packages from hunter on a LAN available server. Packages which we develop internally are hosted on a local instance of GitLab. When adding them to Hunter I create a tag and download this tag from GitLab directly. Without the redirect variable I would have to manually copy the tag to the download server

With this change I can make external Hunter packages available for the build system by putting the archives on the download server. Up until now I had to manually modify Hunter to download from an internal server. This often included making a mirror repository which I also had to update.

I hope this makes it clearer for you

Reply:
> I intend to put the archives of external packages from hunter on a LAN available server. Packages which we develop internally are hosted on a local instance of GitLab

It means you have two download servers. Turn `HUNTER_DOWNLOAD_SERVER` into list. Add two servers to this list. If URL of package doesn't match any of the entry then do substitute procedure (using first element of the list). If URL of package match any server then use original URL. How about that?

Reply:
I like your suggetion, but I'd like to reserve the multiple URLs in `HUNTER_DOWNLOAD_SERVER` for a possible future usage of defining multiple download server to try in sequence until one succeeds

from the 3.10 documentation of [ExternalProject](https://cmake.org/cmake/help/v3.10/module/ExternalProject.html)
> URL Download
>
>    URL <url1> [<url2>...]
>        List of paths and/or URL(s) of the external project’s source. When more than one URL is given, they are tried in turn until one succeeds. A URL may be an ordinary path in the local file system (in which case it must be the only URL provided) or any downloadable URL supported by the file(DOWNLOAD) command. A local filesystem path may refer to either an existing directory or to an archive file, whereas a URL is expected to point to a file which can be treated as an archive. When an archive is used, it will be unpacked automatically unless the DOWNLOAD_NO_EXTRACT option is set to prevent it. The archive type is determined by inspecting the actual content rather than using logic based on the file extension.

Reply:
> but I'd like to reserve the multiple URLs in HUNTER_DOWNLOAD_SERVER for a possible future usage of defining multiple download server to try

I don't see a conflict here.

Reply:
With the single download server and then other servers I do understand your proposal. But with multiple download servers in one variable there could be a problem

assume we have
`HUNTER_DOWNLOAD_SERVER="http://dl-server;https://github.com;https://gitlab.com"`

First the case, where the first element in the list is the download server and the others are (more or less) the `NO_REDIRECT` patterns:
- package foo defines the URL `http://example.com/foo/v1.2.3.tar.gz`. Then the download server kicks in and the URL is redirected to `http://dl-server/foo/v1.2.3/SHASUM/v1.2.3.tar.gz`. This case works very good.
- package bar defines the URL `https://github.com/example/bar/archive/v1.2.3.tar.gz`. Then the URL is not changed and the downloadserver is NOT used (as intended)

Everything works as expected. The elements in the variable `HUNTER_DOWNLOAD_SERVER` are the only servers packages are allowed to be downloaded from. The first element is special, because it is used as base for the URL-creataion

The next case if we say we want multiple download servers defined and each server is tried one after another.
- package foo would be (like in the first case) redirected to `http://dl-server/foo/v1.2.3/SHASUM/v1.2.3.tar.gz`. If this fails the next entry would be tried and the next url would be `https://github.com/foo/v1.2.3/SHASUM/v1.2.3.tar.gz`. This will fail and the next entry would be `https://gitlab.com/foo/v1.2.3/SHASUM/v1.2.3.tar.gz`, which will also fail
- package foo would be downloaded from the original URL, because github.com is in the list

I find the second usage with multiple download server very confusing, because there is no clear distinction between  DOWNLOAD_SERVER and NO_REDIRECT elements

Reply:
if we only have one download server then we could use just one variable `HUNTER_DOWNLOAD_SERVER`. Then this variable would contain a list of all servers hunter would download from, just the first URL in the list is treated in a special way to also be used as base for other URLs not listed in the variable `HUNTER_DOWNLOAD_SERVER`

if we want to be future proof and allow multiple download server, then just one variable is confusing (in my opinion). We should think about how to solve it or if we want to allow multiple download server (in the future)

Reply:
> package foo would be downloaded from the original URL, because github.com is in the list

I don't expect package to be downloaded from original URL if `HUNTER_DOWNLOAD_SERVER` defined. You will retry `http://dl-server/foo/v1.2.3/SHASUM/v1.2.3.tar.gz`.

Reply:
> then just one variable is confusing

Using more that one download server is confusing anyway. I was thinking that `HUNTER_DOWNLOAD_SERVER` will be used as the only server to download sources from. It can be maintained outside, not in Hunter.

Reply:
---

* Package name: `foo`
* Package URL: `https://foo.com/downloads/foo-1.0.tar.gz`

---

* Package name: `boo`
* Package URL: `https://server-2.com/downloads/boo-3.0.tar.gz`

---

If `HUNTER_DOWNLOAD_SERVER` is empty we use URLs `https://foo.com/downloads/foo-1.0.tar.gz` and `https://server-2.com/downloads/boo-3.0.tar.gz` to download sources. Retry logic implemented in `ExternalProject_Add`.

If `HUNTER_DOWNLOAD_SERVER` is a list of servers `https://server-1.com;https://server-2.com;https://server-3.com` we do analyze original package URL. If original URL matches one of the server we leaving it untouched and set as a server with high priority.

For package `foo` we pass next URLs to `ExternalProject_Add` (original URL is not used):
* `https://server-1.com/foo/1.0/SHASUM/foo-1.0.tar.gz`
* `https://server-2.com/foo/1.0/SHASUM/foo-1.0.tar.gz`
* `https://server-3.com/foo/1.0/SHASUM/foo-1.0.tar.gz`

For package `boo` we pass next URLs to `ExternalProject_Add`:
* `https://server-2.com/downloads/boo-3.0.tar.gz` (take priority, original URL used)
* `https://server-1.com/boo/3.0/SHASUM/boo-3.0.tar.gz`
* `https://server-3.com/boo/3.0/SHASUM/boo-3.0.tar.gz`

Reply:
very nice proposal! I like it!

one last question I have is how `external_project_add` of older cmake versions (<3.7) handle a list of URLs instead of a single URL

3.6 documentation https://cmake.org/cmake/help/v3.6/module/ExternalProject.html
> URL /.../src.tgz
>    Full path or URL of source

3.7 documentation https://cmake.org/cmake/help/v3.7/module/ExternalProject.html
> URL /.../src.tgz [/.../src.tgz]...
>    Full path or URL(s) of source. Multiple URLs are allowed as mirrors.

Reply:
> one last question I have is how external_project_add of older cmake versions (<3.7) handle a list of URLs instead of a single URL

You can check `CMAKE_VERSION` and allow multiple URLs only for CMake 3.7+. Mention in here:
* https://docs.hunter.sh/en/latest/quick-start/cmake.html

Reply:
Anyway let's finish the first part without multiple URLs feature. You can add it in next pull request.